tempesta_fw/http2/huffman.c

```diff
@@ -0,0 +1,641 @@
+			   const HTState * __restrict state)
+{
+	ufast i;
+	for (;;) {
+		fast shift;
+		if (Opt_Unlikely(current == -HT_NBITS)) {
+			if (Opt_Likely(state == HTDecode)) {
+				return 0;
+			}
+			else {
+				return HTTP2Error_Huffman_CodeTooShort;
+			}
+		}
+		i = (c << -current) & HT_NMASK;
+		shift = state[i].shift;
+		if (shift >= 0) {
+			if (shift > current + HT_NBITS) {
+				break;
+			}
+			* dst++ = (char) state[i].offset;
+			current -= shift;
+			state = HTDecode;
+		}
+		else {
+			/*
+			 * Last full prefix also processed here, to allow
+			 * EOS padding detection. Condition here equivalent to
+			 * the "-shift >= current + HT_NBITS", but working faster:
+			 */
+			if (Opt_Likely(shift <= -HT_NBITS - current)) {
+				break;
+			}
+			return state[i].offset == 0 ? HTTP2Error_Huffman_UnexpectedEOS :
+						      HTTP2Error_Huffman_InvalidCode;
+		}
+	}
+	if (Opt_Likely(state == HTDecode &&
+	       (i ^ (HT_EOS_HIGH >> 1)) < (1U << -current)))
+	{
+		return 0;
+	}
+	else {
+		return HTTP2Error_Huffman_CodeTooShort;
+	}
+}
+
+static fast
+http2_huffman_decode_tail_s (ufast			c,
+			     char	   * __restrict dst,
+			     fast			current,
+			     const HTState * __restrict state)
+{
+	int16 offset;
+	fast shift;
+	ufast i;
+	if (Opt_Unlikely(current == -HT_MBITS)) {
+		return HTTP2Error_Huffman_CodeTooShort;
+	}
+	i = (c << -current) & HT_MMASK;
+	shift = state[i].shift;
+	offset = state[i].offset;
+	if (Opt_Likely(shift >= 0)) {
+		if (Opt_Unlikely(shift > current + HT_NBITS)) {
+			return HTTP2Error_Huffman_CodeTooShort;
+		}
+		* dst++ = (char) offset;
+		current -= shift;
+		return http2_huffman_decode_tail(c, dst, current, HTDecode);
+	}
+	else {
+		/*
+		 * Condition here equivalent to the "-shift > current + HT_NBITS",
+		 * but working faster:
+		 */
+		if (Opt_Unlikely(shift < -HT_NBITS - current)) {
+			return HTTP2Error_Huffman_CodeTooShort;
+		}
+		return offset == 0 ? HTTP2Error_Huffman_UnexpectedEOS :
+				     HTTP2Error_Huffman_InvalidCode;
+	}
+}
+
+fast
+http2_huffman_decode (const char * __restrict source,
+			    char * __restrict dst,
+			    uwide	      n)
+{
+	if (n) {
+		const uchar * __restrict src = (const uchar *) source;
+		ufast c = * src++;
+		fast current = 8 - HT_NBITS;
+		int16 offset;
+		n--;
+		for (;;) {
+			const HTState * __restrict state;
+Root:			state = HTDecode;
+			for (;;) {
+				fast shift;
+				ufast i;
+				if (current <= 0) {
+					if (Opt_Likely(n)) {
+						current += 8;
+						c = Bit_Join8(c, * src++);
+						n--;
+					}
+					else {
+					     /* Last full prefix also processed here */
+					     /* (see current <= 0 above): */
+						return http2_huffman_decode_tail(
+							c, dst, current, state
+						);
+					}
+				}
+				i = (c >> current) & HT_NMASK;
+				shift = state[i].shift;
+				offset = state[i].offset;
+				if (shift >= 0) {
+					* dst++ = (char) offset;
+					current -= shift;
+					goto Root;
+				}
+				else {
+					current += shift;
+				#ifdef HT_BALANCED_TREE
```

Comment:
I believe the code will be just dead for near future. Meantime we have to review and support the code, so we do not maintain unused code. Moreover, we do not develop code for "possible future needs": we don't know the future until it happens, so requirements and an environment can change. In general, we develop code when it's needed only.